cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppStateHandler.java

```diff
@@ -0,0 +1,184 @@
+    applicationLifecycleService.saveState(appStateRequest);
+    responder.sendStatus(HttpResponseStatus.OK);
+  }
+
+  /**
+   * Delete a {@link AppState} for a given app-name and state-key.
+   */
+  @DELETE
+  @Path("/namespaces/{namespace-id}/apps/{app-name}/states/{state-key}")
+  public void deleteState(HttpRequest request, HttpResponder responder,
+                          @PathParam("namespace-id") String namespaceId,
+                          @PathParam("app-name") String appName,
+                          @PathParam("state-key") String stateKey) throws Exception {
+    validateInput(namespaceId, appName, stateKey);
+    AppState appStateRequest = new AppState(namespaceId, appName, stateKey);
+    applicationLifecycleService.deleteState(appStateRequest);
+    responder.sendStatus(HttpResponseStatus.OK);
+  }
+
+  /**
+   * Delete all {@link AppState} for a given app-name.
+   */
+  @DELETE
+  @Path("/namespaces/{namespace-id}/apps/{app-name}/states")
+  public void deleteAllStates(HttpRequest request, HttpResponder responder,
+                              @PathParam("namespace-id") String namespaceId,
+                              @PathParam("app-name") String appName) throws Exception {
+    validateInput(namespaceId, appName);
+    AppState appStateRequest = new AppState(namespaceId, appName);
+    applicationLifecycleService.deleteAllStates(appStateRequest);
+    responder.sendStatus(HttpResponseStatus.OK);
+  }
+
+  private void validateInput(String namespace, String appName) throws Exception {
+    NamespaceId namespaceId = validateNamespace(namespace);
+
+    validateApplication(appName, namespaceId);
+  }
+
+  private void validateInput(String namespace, String appName, String stateKey) throws Exception {
+    validateInput(namespace, appName);
+
+    validateStateKey(stateKey);
+  }
+
+  private void validateInput(String namespace, String appName, String stateKey, byte[] stateValue) throws Exception {
+    validateInput(namespace, appName, stateKey);
+
+    validateStateValue(stateValue);
+  }
+
+  private void validateApplication(String appName, NamespaceId namespaceId)
+    throws BadRequestException, ApplicationNotFoundException {
+    if (appName == null) {
+      throw new BadRequestException("Path parameter app-id cannot be empty");
+    }
+
+    ApplicationId appId = namespaceId.app(appName);
+    applicationLifecycleService.validateApplication(appId);
+  }
+
+  private void validateStateKey(String stateKey) throws BadRequestException {
+    if (stateKey == null) {
+      throw new BadRequestException("Path parameter state-key cannot be empty");
+    }
+  }
+
+  private void validateStateValue(byte[] stateValue) throws BadRequestException {
+    if (stateValue == null) {
+      throw new BadRequestException("Path parameter state-value cannot be empty");
+    }
+  }
+
+  private NamespaceId validateNamespace(String namespace) throws Exception {
```

Comment:
instead of throws Exception, explicitly declare each type of exception that can be thrown.